AGENT-1250: Simplified cluster operations for disconnected environments

andfasano opened the pull request:
This patch contains the enhancement proposal for the simplified cluster operations in the disconnected environments.

andfasano committed:
enhancement proposal for the simplified cluster operations

zaneb commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+
+The [distribution/distribution](https://github.com/distribution/distribution)
+registry will be used to run the registry services on each control plane
+node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)
```
Included in, and built from the same source as, so that CVEs are handled without any additional overhead.

zaneb commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+* Offer a simplified installation experience for disconnected environments by
+  leveraging existing tools like the Agent-Based Installer, Assisted Installer
+  and OpenShift Appliance
+* Allow the user to optionally install/upgrade a predefined subset of OLM
```
Since this is not currently covered in this enhancement (and will probably be a Whole Thing), maybe we should move this goal to a separate enhancement.

andfasano replied:
Ok

zaneb commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+  was always available in the environment. Removing such dependency across
+  all the possible locations appeared to be particularly complex and
+  challenging (and moreover no formal way to enforce it once removed)
+
```
I'd say all 3 of the alternatives from the discussion doc are worth documenting as rejected alternatives:

* the initial version of this proposal, before we realised that we would have to expose the local registries via api-int.
* the version using PinnedImageSet
* the idea of writing a registry that exposes the containers-storage as a registry

zaneb replied:
Yeah, I have chatted to Colin about that. Unfortunately there are some [fundamental limitations to what is standardised](https://github.com/zaneb/distribution-containers-storage?tab=readme-ov-file#when-this-does-not-work) in the OCI Distribution protocol that we concluded made this approach too high-risk for our purposes.

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
````diff
@@ -0,0 +1,425 @@
+This is an example of how the CR will look like:
+
+```
+apiVersion: releases.openshift.io/v1
````
APIs should start life as v1alpha1 and then, when we are confident in promoting the feature, go through a new round of review to implement any feedback about the API, before moving to v1

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+apiVersion: releases.openshift.io/v1
+kind: InternalReleaseImage
+metadata:
+  name: openshift-releases
```
Our singletons are usually called `cluster`, I expect this one should be the same

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+metadata:
+  name: openshift-releases
+spec:
+  releases:
```
So as an end user, how do I use this API? I create this resource and I specify the images that are available? Don't I need to tell something where to load the images from?

JoelSpeed replied:
How do I remove an image? If I delete it from this list, will it be removed from the local registry and then the status field?

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+  releases:
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
+    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
+  enabled: true
```
Why do you need `enabled`? Surely the presence of the resource singleton implies that the feature is enabled?

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
````diff
@@ -0,0 +1,425 @@
+  ...
+```
+
+* The `spec.releases` will be used to configure the desired release contents.
+  Adding (or removing) an entry from the list will trigger an update on all the
+  control plane nodes.
````
Where do the control plane nodes source the images from?

JoelSpeed replied:
I think I meant, how does the control plane know where to copy the images from, I didn't see anywhere in the API where you configure "ISO is loaded to this mount point"

sdodson replied:
This wasn't clear to me either.

andfasano committed:
fixed filename and reviewers list

andfasano committed:
some initial doc reviews

andfasano commented:
(sorry, the file rename due the markdownlint failures scrambled a little bit the latest comments, I will keep addressing the unresolved one in the next commits - hopefully no more file renaming :disappointed_relieved: )

danielerez commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+the user could start the usual offline upgrade process via the `oc adm upgrade
+--allow-explicit-upgrade --to-image=<release-image>`.
+At the end of the upgrade the user may decide to delete the previous release
+entry from the IRI resource. In such case the MCD on each control plane node
```
What is the API for removing a previous release?
The suggestion is that MCD will monitor IRI and handle the removal?
I.e. IRI stores the list of images, and MCD removes every image?

danielerez commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+won't be sufficient to support the joining process. For that, it will be
+necessary to run the registry on a host network port accessible from outside
+the cluster. In addition to that, the registry will be secured via a
+pull-secret. Users will be required to add this port to the api-int load
```
The port should be configurable?
Or should we just guide users the open some constant port (e.g. 5000)?

danielerez commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+
+![alt install phase](simplified-operations-install.png)
+
+A systemd service injected by MCO will ensure the registry will be available
```
Same as the current approach in the appliance ([start-local-registry.service](https://github.com/openshift/appliance/blob/master/data/services/common/start-local-registry.service#L1))?

danielerez commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+
+#### Upgrade
+
+A udev rule, along with a systemd service injected by MCO will detect when an
```
Something like the current [udev](https://github.com/openshift/appliance/blob/master/data/udev/rules.d/99-upgrade-iso.rules#L1) rule in the appliance, right?

danielerez commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+
+### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
```
Hmm, do we know how to test it e2e?
E.g. use a config-image instead of the interactive 'below-sea-level' part?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+to each local control plane node registry), managed by the Machine Config
+Daemon (MCD). Once the copy has been completed on all the control plane nodes,
+the user could start the usual offline upgrade process via the `oc adm upgrade
+--allow-explicit-upgrade --to-image=<release-image>`.
```
What is the suggested approach to extract the relevant 'release-image' from the ISO?
Or maybe since the ISO is constructed for a specific release image anyway, we can keep it transparent for users? E.g. apply a MachineConfig that starts the upgrade (and detects the right version internally)

sdodson replied:
I've lost track of the specific validations the CVO does but I believe the way it's intended to work today is if you provide a symbolic version it has to be on the graph. As soon as you provide a specific pull spec (--allow-explicit-upgrade) the CVO assumes you know what you're doing, not that that's ever been documented explicitly. There's a lot of blurred lines between CLI side guards and what the API accepts.

@wking can you check me here? what protections are in place here and what do you think would be best way to keep them from making a mistake by potentially selecting the wrong mirrored image when expecting to upgrade? Maybe it's time to move all the client side protections into the ClusterVersion CR and CVO via a new interface?

zaneb replied:
It would be great if we had symbolic versions. What I think we'd need to do that is a mapping of tags to sha digests that is cryptographically signed. If that existed it would be fairly easy to build into the ISO, but the infrastructure to set it up if it doesn't would be potentially daunting. Do you know of any existing artifacts that could act as building blocks here? A brief reading of sigstore docs suggests that it won't help.

palonsoro replied:
FWIW: Before OSUS (and oc-mirror) were around, when performing OCP4 upgrades on disconnected environments, they had to be done with `--allow-explicit-upgrade` and that did have the risks you are pointing out here of not properly obeying the graph. In that era, the message used to be that you had to be careful to follow the graph on your own or you risked to not be properly supported.

However, I'd prefer to have safeguard mechanisms like the ones you are considering.

zaneb replied:
Arguably there's a high probability that most disconnected upgrades still follow this path and don't use OSUS, and that we should therefore punt this to a future enhancement rather than block on it now. (I agree that I'd _prefer_ to have it, all things being equal.)

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+the user could start the usual offline upgrade process via the `oc adm upgrade
+--allow-explicit-upgrade --to-image=<release-image>`.
+At the end of the upgrade the user may decide to delete the previous release
+entry from the IRI resource. In such case the MCD on each control plane node
```
What is the API for removing a previous release?
The suggestion is that MCD will monitor IRI and handle the removal?
I.e. IRI stores the list of images, and MCD removes every image?
If so, how will we provide the full list of images? Include the oc-mirror mapping in ISO?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+won't be sufficient to support the joining process. For that, it will be
+necessary to run the registry on a host network port accessible from outside
+the cluster. In addition to that, the registry will be secured via a
+pull-secret. Users will be required to add this port to the api-int load
```
The port should be configurable?
Or should we just guide users the open some constant port (e.g. 5000)?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+The [distribution/distribution](https://github.com/distribution/distribution)
+registry will be used to run the registry services on each control plane
+node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)
+repo image, so that it will be available within an OCP release image payload.
```
Which image in OCP payload will include the registry?
Will it be addressed as part of this effort, or do we have another enhancement for this?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+![alt bootstrap phase](simplified-operations-bootstrap.png)
+
+Initially the registry image will be placed into a read-only [additional image
```
Currently we simply use 'podman load' on the registry tar file.
Are you referring to the same flow?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+![alt install phase](simplified-operations-install.png)
+
+A systemd service injected by MCO will ensure the registry will be available
```
Same as the current approach in the appliance ([start-local-registry.service](https://github.com/openshift/appliance/blob/master/data/services/common/start-local-registry.service#L1))?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+#### Upgrade
+
+A udev rule, along with a systemd service injected by MCO will detect when an
```
Something like the current [udev](https://github.com/openshift/appliance/blob/master/data/udev/rules.d/99-upgrade-iso.rules#L1) rule in the appliance, right?

danielerez commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
```
Hmm, do we know how to test it e2e?
E.g. use a config-image instead of the interactive 'below-sea-level' part?

djoshy commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
+    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
+  enabled: true
+status:
```
Would per-node status reporting be beneficial here? I would worry about multiple MCDs acting on a singular resource and trying to update the same status field causing conflicts.

zaneb replied:
If the CR is the source of truth, then per-node status reporting is essential.
However, if the controller can always ask MCD what the current status is then it could roll those up to a single status in the CR.
The enhancement for PinnedImageSet says that per-node reporting via the `MachineConfigNode` CRs will be "considered". I don't know what was actually implemented in the end.

yuqi-zhang replied:
It exists as a subfield on the MCN status: https://github.com/openshift/api/blob/master/machineconfiguration/v1/types_machineconfignode.go#L170

zaneb replied:
Thanks! Probably makes sense to do the same then I would think.

andfasano replied:
Will prepare a first proposal

djoshy commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+At this point the user could edit the existing IRI resource to add the new
+release to trigger the copy process (from the temporary registry
+to each local control plane node registry), managed by the Machine Config
+Daemon (MCD). Once the copy has been completed on all the control plane nodes,
```
It might be helpful to have a singular controller for orchestrating the copy process across the MCDs running on the control plane nodes. This makes it easier to do aggregate several node states into overall conditions and to report error modes at the operator level.

zaneb replied:
This design was inspired by [PinnedImageSet](https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/pin-and-pre-load-images.md) in the hope that much of that can be... if not _reused_, at least used as a template.
IIUC PinnedImageSet does have its own controller running inside the MCO. I don't think the language here was intended to exclude that, so it could probably be clarified slightly.

djoshy replied:
Ah, I see. Agreed that we should clarify if PIS is going to be reused here. 

They're specified at the [MachineConfigPool](https://github.com/openshift/api/blob/master/machineconfiguration/v1/types.go#L454) level and the node level images are represented in the status of the [MachineConfigNode](https://github.com/openshift/api/blob/master/machineconfiguration/v1/types_machineconfignode.go#L170) objects. We should be able to leverage that.  We'd need some sort of translation between InternalReleaseImage <=> PinnedImageSets; maybe at the control-plane level? 

I'm also not sure if PIS currently works in bootstrap/firstboot mode, so that might be something I can try getting clarification from the team on(cc @yuqi-zhang if you have context here). The flow above hints that IRI manifest is added at the end of the installation, so perhaps the MCO translation piece is not relevant at bootstrapping time?

yuqi-zhang replied:
Hmm, I guess I'm not too clear on this either. It reads to me that the install flow is completely handled by the agent installer (i.e. it will "set up" the registry and image content from the install iso on all the control plane nodes), and upgrade wise, that responsibility falls onto the MCD via specifications in IRI? Does IRI come into the install process at all through the worker nodes?

Re: upgrade, my read of the process outlined below is that:
1. the user plugs in a usb disk ("attach the extended ISO") to the first control plane node
2. the user writes IRI for that specific node, so that the MCD copies it
3. the user unplugs it, plugs it into the second control plane node
4. write IRI to the second node
repeat until done for control plane, then trigger the upgrade for the whole cluster, where the 3 control plane node reads off itself and then have the workers read off of the control planes?

If that's the case, I think regardless we'd need some daemon process to handle the actual "copy". We probably also need some controller handling the central state though,  to ensure the user has successfully completed all copies and write some status before they can upgrade.

zaneb replied:
> It would certainly be advantageous if we could rely on the IRI and MCO even at install time to create the local registry

On reflection, this wouldn't work for single-node :slightly_frowning_face:

djoshy commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+* The `spec.releases` will be used to configure the desired release contents.
+  Adding (or removing) an entry from the list will trigger an update on all the
+  control plane nodes.
+* The `spec.enabled` will be used to turn on / off the feature. In particular,
```
Would be good to have a small blurb about `status.enabled` here too.

JoelSpeed replied:
In that case, I would assume that customers will test the change on a staging environment right? If the new registry doesn't work out, will the existing images cached in the CRI still work? I guess unless they have `imagePullPolicy: Always`?

What would you expect time to recovery to be for rebuilding the local registry once the IRI is re-instated?

andfasano replied:
> Do we actually expect users to frequently disable and then re-enable this? In https://github.com/openshift/enhancements/pull/1821#discussion_r2278399494 the suggestion is that it will be disabled for good?

ATM I do not expect any frequent feature opt-in/out (if not by mistake, because the user inadvertently removed the IRI resource), so I agree to delete everything when the resource is removed.

zaneb replied:
Just like we won't let people delete the current release image (the one that CVO is targeting or upgrading from) from the image list, we also won't let people disable the whole feature (by deleting the IRI CR) under the same circumstances.
So in order to disable the feature, the user will have to use CVO to upgrade (or 'upgrade') to a release in the external registry _first_. After that we can safely disable the feature and delete any images stored on disk.
(We should document this explicitly.)

zaneb replied:
[Yes](https://github.com/openshift/enhancements/pull/1821#discussion_r2282081134), that's what we decided.

djoshy commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+extended RHCOS ISO is attached, an then mount it to launch a temporary registry
+to serve its content.
+After the user modified the IRI resource with the new release pullspec, the new
+OCP release image payload will be copied on each node registry using skopeo.
```
Perhaps a silly question, but is there value in these images being verified for integrity? For example, say the node goes through a reboot and the new MCD comes up and examines the IRI resource. How does it know the releases mentioned by the IRI spec field matches what is in its local registry?

zaneb replied:
Not a silly question!
Ideally, all images are referenced by digest rather than tag. That's enough to maintain integrity.
However, there's an obvious usability gap there, and how we bridge it is underspecified by this proposal.

zaneb replied:
I think we need to document each step from the perspective of the user:

* An ISO is attached, and a registry is made available
* The user has to explicitly acknowledge that they intended to do that by putting some data into the IRI resource. What data is that?
  * A version number? How will the system verify that the image it is pulling is the one that the user is explicitly acknowledging?
  * A pullspec with a tag? How will the system verify that the image it is pulling is the one that the user is explicitly acknowledging?
  * A pullspec with a digest? Where will the user copy the pullspec from? (And how will they tell which one to remove later when they are done with it?)
* The user then initiates an update with CVO. What pullspec will they use as the target? Where will the user copy the pullspec from? Is there any way to verify the [update graph](https://github.com/openshift/enhancements/pull/1821#discussion_r2278096071)?
  * The original quay pullspec with a tag? (I don't think this works because IDMS won't handle it and we don't want to use ITMS.)
  * The original quay pullspec with a digest?
  * A pullspec pointing directly to the mirror with a tag?
  * A pullspec pointing directly to the mirror with a digest?

palonsoro reviewed (commented):
I'd have some questions and I also found a typo.

zaneb commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+The [distribution/distribution](https://github.com/distribution/distribution)
+registry will be used to run the registry services on each control plane
+node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)
```
`openshift/image-registry` is a downstream fork, and is part of the release payload. So it will be updated along with the release payload.
There is some extra complexity because this image is actually hosting the registry, so the image also needs to be obtained from the release payload and put into a containers/storage store where it is directly accessible by podman.

zaneb replied:
Actually I was a bit inaccurate there. It's not a fork but a wrapper that vendors in the upstream distribution/distribution.
The [plan](https://github.com/openshift/image-registry/pull/426) is to add another binary to the container that is just building distribution/distribution from the vendor directory without any OpenShift-specific bits added.
That way any CVE fixes we apply automatically get included for both this purpose and the image-registry-operator purposes.

oceanc80 commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+From a very high-level perspective, this proposal consists of building and
+releasing an extended RHCOS live ISO that includes a full OCP release payload
+(with a selected subset of OLM operators) and all the necessary
```
Wanted to clarify my understanding, is the scope of this work to include a selected subset of OLM operators in the release payload but for now not allow them to be installed or upgraded? Is there an eventual goal to integrate this with OLM, eg an expectation that OLM would automatically pick up this registry on start?

yuqi-zhang commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+  enabled: true
+status:
+  releases:
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
```
Thinking about what we should report here, I wonder if each individual status entry for each payload should have subfields around progressing/success/error/etc, instead of overall conditions, so it's easier to see which payloads are there, which are being pulled, etc.

Although since a user is unlikely to specify more than 2 payloads here at a time, perhaps we just report the most recent error? Would the MCD would be constantly reconciling all releases it expects to be on disk? Or just the latest added/removed one?

yuqi-zhang replied:
In that case I think the status should have sub sections per release on the status. i.e. it would be nice to see how many nodes have successfully pulled each release, what errors there are, etc.

yuqi-zhang commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+ISO in the storage format used by the [distribution/distribution](https://github.com/distribution/distribution)
+registry’s filesystem storage plugin.
+
+#### Distribution/distribution registry OCP integration
```
The image registry here has some interesting overlap with the MCO's image mode efforts, namely:
1. the design seems to be a lightweight, cluster-only registry used for cluster operations that we manage (somewhat through the MCO)
2. this registry could potentially be used for os image build distribution for this cluster

cc @pablintino who has been investigating image registry use cases for image mode

JoelSpeed commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+  control plane nodes.
+* The `spec.enabled` will be used to turn on / off the feature. In particular,
+  when turned off, all the registry services running on the control plane nodes
+  will be stopped.
```
Do you think folks will often want to be able to disable the serving of the content locally without clearing the disks? When would I want to do that?

I kind of thought this feature would be something users keep on once it is enabled and that the reasons for turning it off again would be minimal?

palonsoro replied:
The only possible "realistic" use case to turn the feature of that I can think of is if the cluster was installed via this method, but later the customer decides to install a corporate registry and switch OCP to use it.

Same way, it might be possible that a user is using a mirror registry, yet they would decide to deprecate it and enable this feature day-2.

In both cases, it doesn't look like we may need to keep the releases on disk wasting space, specially when getting them back should be as easy as attaching an ISO, right?

zaneb replied:
I can imagine people turning it off for the reason Pablo suggested, but I also agree that disabling it should remove all of the content on disk.

sdodson commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+#### Upgrade
+
+A udev rule, along with a systemd service injected by MCO will detect when an
+extended RHCOS ISO is attached, an then mount it to launch a temporary registry
+to serve its content.
+After the user modified the IRI resource with the new release pullspec, the new
+OCP release image payload will be copied on each node registry using skopeo.
```
Skopeo is pointed directly at the host service or does it also get an IMDS entry and skopeo thinks it's talking to quay.io? If it had an IMDS entry could the IRI controller be used, unaltered, in a connected environment because it would eventually synchronize from the origin?

zaneb replied:
There's something to be said for that, and also something to be said for not doing that because users who are connected but bandwidth-constrained will hate us if we download data over the network when they already supplied it locally :smile:

Either way, the answer needs to be made explicit in this proposal.

sdodson replied:
They'd have supplied it locally via an IMDS pointed at their local registry, right? I assume there's a way to say never even attempt to reach the origin and fail instead?

zaneb commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+      lastTransitionTime: "2025-07-12T14:02:00Z"
+      reason: "ReleasesSynchronized"
+      message: "The previously managed releases are available."
+    - type: "Progressing"
+      status: "True"
+      lastTransitionTime: "2025-07-12T13:13:00Z"
+      reason: "AddingRelease"
+      message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."
```
I think adding and removing a release at the same time is fine. The limitation will be that you can't remove a release that the CVO is currently using (i.e. is the current target or the one we are in the process of upgrading from).

zaneb replied:
That would be the easiest way.
An alternative would be to say that we will independently look up what the current release image is, and if it appears in the Status do not remove it from the registry even if it has disappeared from the Spec.

The former is more explicit and therefore easier to explain to users I think.

palonsoro replied:
Yep, I agree. If we silently keep the current release in the status when removed from spec, it would become necessary to constantly watch if it is still the current release or not and eventually schedule the removal. More complex.

I agree that just rejecting the update to the spec is easier.

zaneb commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+The [distribution/distribution](https://github.com/distribution/distribution)
+registry will be used to run the registry services on each control plane
+node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)
```
> I guess the assumption here is that you'd also have a k8s service to dispatch to those?

I believe what we're saying is that we'll expose it on a host networking port, and it will be accessible via the api-int load balancer (in the case of platform:none we'll require users to add that port to their load balancer).

palonsoro commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+1. The user downloads the extended RHCOS ISO from the Red Hat Customer Portal,
+   by selecting a compatible newer version release for the target cluster.
+2. The user moves the extended RHCOS ISO into the disconnected environment.
+3. The user attaches the extended RHCOS ISO to one of the control plane nodes
+   of the target cluster.
```
On one hand, how/where the image is exposed might vary depending on the environment (and the details of the driver used for the virtual or real drive involved).

On the other hand, it would be ideal to abstract the user as much as possible of this kind of host-level details.

So ideally this should be detected by just attaching it (without requiring the user to mount anything manually), yet it might be wise to keep some way for manual override (in case the customer has the cluster in an environment where the ISO is attached in a way that cannot be auto-detected).

zaneb replied:
We will label the volume in the ISO. A udev rule will detect this volume label and automatically trigger a systemd service parameterised with the volume mount point. The user doesn't need to know anything about anything. We have used this approach successfully in the past.

JoelSpeed replied:
Is this method documented in this EP or any other EP? Might be worth a short note here to explain it

andfasano replied:
I don't think there's any EP related to that. I will add more details below in the implementation details [Upgrade](https://github.com/openshift/enhancements/pull/1821/files#diff-7b20b50dcbb0b982828fe87930f63c142457b54311f85d0a73468c86ee2165c1R344) paragraph, where it is mentioned

zaneb commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+#### Post-upgrade optional steps
+
+1. The user edits the IRI resource by removing the release image pullspec not
+   in use anymore.
```
Currently ~40GiB per release with all the OLM operators, but we have more OLM operators to add, so probably closer to 60. For just OCP, about 20.

JoelSpeed replied:
We need to be able to store at least 2 full payload copies right? To be able to handle current + upgrade?

In which case documenting that at least 120 GiB of storage is needed on each node seems quite important. I believe across most platforms we recommend 128GiB or less for the default volume size so this extra space is something to consider

palonsoro replied:
Indeed. It is important that we properly explain that the ~120GB (or whatever) is required for the releases and in addition to whatever disk space may be also required by workloads, etc.

Otherwise, some users may just install hard drives of 130GB, make them full with the 2 releases and complain when they cannot start 50 containers that do heavy usage of `EmptyDir volumes

andfasano replied:
Ok, I'll add an explicit node for the documentation (and for the live Assisted Service pre-flight check) related to the increased storage size required

zaneb replied:
I don't disagree with any of this, but note that we're only storing the images on the control plane nodes.

JoelSpeed replied:
I don't see anything in the body of the EP currently documenting the additional disk overhead for control plane nodes, lets get something added and we can resolve this thread

andfasano replied:
Added

andfasano committed:
Update enhancements/agent-installer/simplified-registry-less-cluster-operations.md

Co-authored-by: Pablo Alonso Rodriguez <palonsoro@users.noreply.github.com>

andfasano committed:
swapped IRI spec.enable field with resource deletion

zaneb commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+  adopted by the Machine Config Server (MCS).
+  This is a new requirement that will need to be properly documented. In
+  addition, to minimize the risks, the registry will have to be secured
+  via a pull-secret.
```
We'll probably need to document how the user can force the credentials to be rotated.
I imagine that at installation time we'll generate credentials and save them in a Secret, as well as add the auth data to the cluster's pull secret.
If the user wants to rotate they'll need to delete the Secret (MCO will automatically replace?) and manually update the cluster pull secret.

JoelSpeed replied:
This sounds like an important workflow to document in the EP

andfasano replied:
Added some additional details for this part.

andfasano committed:
refined release and resource removal process

andfasano committed:
modified IRI to specify a list of object releases

Currently, only the name field is specified for each release

JoelSpeed commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,426 @@
+
+#### Upgrade
+
+A udev rule, along with a systemd service injected by MCO will detect when an
```
Possibly both, but since the MCD is going to be looking at this (and is the closest openshift code to the mount), would it not make sense for it to be reported in the MCN (effectively each MCD gets an MCN to write information it knows about to)

andfasano committed:
Fixed spec.currentRelease field

andfasano commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,461 @@
+by the IRI itself. To opt-out from the feature, the user will have to setup
+his/her own registry with the required mirrored content and perform a cluster
+upgrade using it.
+Once successfully completed, the `InternalReleaseImageController` will 
+blank out the `spec.currentRelease` field to indicate that none of the its
+managed release payloads are currently being used for the cluster. At this
+point it would be possible for the user to remove the IRI resource.
```
As per the previous [comment](https://github.com/openshift/enhancements/pull/1821#discussion_r2314237718) if VAP doesn't support cross-reference check this would be limited to the controller level

JoelSpeed replied:
See my response there, you can absolutely achieve this using a VAP

JoelSpeed commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
````diff
@@ -0,0 +1,461 @@
+  ...
+```
+
+* The `spec.releases` contains a list of release objects, and it will be used
+  to configure the desired release contents. Adding (or removing) an entry from
+  the list will trigger an update on all the control plane nodes.
````
That would be my assumption yes, lets add a note that when this happens it will be reported (and a suggestion of what that status might look like would be good) please

JoelSpeed commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,461 @@
+  releases:
+    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
+    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
+  currentRelease:
```
> So the list must be able to accommodate at least 3 items.

In the future, EUS upgrades may span larger drifts than even that, we can limit this to 3 now and expand later, but we should consider that this may need to be larger and the implications of that

Worth pulling in OTA for their thoughts on how many they might expect to actually be in use on one cluster at a time?

andfasano committed:
review VAP usage for IRI update/deletion

JoelSpeed commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,459 @@
+updating the resource status.
+
+A new `InternalReleaseImageManager` manager will be added to
+_machine-config-daemon_ to handle per-node specific operations.
```
What exactly would the spec be? What information are you expecting to pass to the MCD? (New annotation based APIs are generally forbidden, especially when we control the types)

andfasano replied:
ATM (just) the list of the releases (to be copied from the ISO). In this way when a user edit the IRI resource to add/remove an entry in the release list, the IRI controller could start orchestrating the download/copy process on each node by updating the related MCN resource

andfasano committed:
added a new section for the storage requirements

andfasano commented on enhancements/agent-installer/simplified_cluster_operations.md:
```diff
@@ -0,0 +1,425 @@
+To allow pulling images from the cluster, an ImageDigestMirrorSet
+(IDMS) policy will be added to the cluster manifests - using the `api-int` 
+DNS entry - so that the set of registries running within
+the control plane nodes will become the nominal pullspec of the current release
```
Ok, added them and slightly modify the paragraph (for the IDMS usage)

andfasano committed:
added goal to opt-out from the feature. Reviewed IDMS paragraph

andfasano committed:
updated image-registry inclusion. Added MCN status proposal

andfasano committed:
added release name specification

andfasano committed:
minor fixes

joelanford commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,522 @@
+
+### Non-Goals
+
+* Allow the user to install/upgrade any OLM operators
```
I'm a little confused. I see user stories that talk about optionally installing/upgrading OLM operators, but then this is listed as a non-goal?

joelanford replied:
Ah, maybe `s/any/arbitrary/` ?

joelanford commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,522 @@
+a command line utility for building a disk image to orchestrate an OpenShift
+installation. The Appliance builder will be enhanced with a new 
+`build live-iso` command capable to generate an ISO artifact. 
+The extend ISO builder will be integrated within the official Red Hat build and
+release pipeline, so that a new extended RHCOS ISO could be published following
```
Will this official ISO include OLM operators? If so, which ones? And how do we choose? And of the ones we chose, are we mirroring all channels, bundles, and related images for those operators? For an official ISO, I think we would need to include the all channels, bundles, and related images to ensure that all possible upgrade paths remain intact. But that will bloat the overall ISO size.

Also, OLM catalogs have been intentionally decoupled from the release cadence of the OCP payload (customers can receive updates to their operators without upgrading their cluster). An official image that contains both an OCP payload and a snapshot of OLM operators implies that users of this method need to upgrade their cluster in order to upgrade their OLM operators. Should we have separate ISOs to keep these decoupled?
- One for payload (generated when new payload is released)
- One for OLM catalogs (generated when new catalogs are released, or on some periodic basis)

andfasano replied:
Yes, the ISO will also include a subset of OLM operators. Including all the operators wouldn't be feasible due the huge size of the final artifact.  Even though the scope of the current proposal does not focus on a specific product - but more on the generic mechanism, it is related to https://issues.redhat.com/browse/VIRTSTRAT-60 (as reported initially) where you could find a more complete list of the included operators.

joelanford commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,522 @@
+Finally, a new `InternalReleaseImage` custom resource, managed by MCO, will
+also be added as a cluster manifest to keep track and manage the currently
+stored internal release image.
+
```
Add details of the OLM aspect of the installation here?

andfasano replied:
Do you maybe mean reporting the list of what was exactly included in the current release bundle? If so I will sooner update the proposal with more details on this topic (see also this [comment](https://github.com/openshift/enhancements/pull/1821/files#r2342887223))

andfasano replied:
cc @danmanor for additional comments

joelanford commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,522 @@
+
+## Summary
+
+This enhancement proposal outlines an opinionated workflow to simplify
+on-premises cluster operations - such as installation, upgrade and node
+expansion - by eliminating the need to setup and configure an external
+registry mirror. This approach is specifically designed for disconnected
+environments and leverages existing tools such as Agent-based Installer,
+Assisted Installer and OpenShift Appliance.
```
I think I understand that this approach is simpler because it doesn't involve users setting up mirror registries on their disconnected network, so are we planning to funnel _all_ disconnected users to this approach?

If not, why would a customer need to stick with the traditional oc-mirror approach?
Or if so, should oc-mirror be able to directly create the ISO (and associated cluster objects) to avoid fragmentation of tooling, docs, etc.?

andfasano commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,522 @@
+At the end of the upgrade the user may decide to delete the previous release
+entry from the IRI resource. In such case the MCD on each control plane node
+will take care of removing the older release payload from the local registry.
+
```
This is definitely a good point and something we'd like to elaborate further, since AFAIU the upgrade path for an OLM operator may be more complicated - so any feedback / suggestion is welcomed.

andfasano committed:
minor non-goals refinement

andfasano committed:
minor fixes

andfasano committed:
fixed Proposal section indentation to improve readability

andfasano committed:
added the registry bundle image

Added a more detailed description for its usage, and proposed a new release identifier (automatically discovered) to be used for adding a new release bundle

pablintino reviewed (commented):
Adding some minor comments. Tomorrow I'll give this a second read, but I think it's OK from a MCO point of view.

yuqi-zhang reviewed (commented):
Some more questions inline

andfasano committed:
minor additional clarifications

andfasano commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,571 @@
+[Assisted Service](https://github.com/openshift/assisted-service/)
+and [Assisted Installer](https://github.com/openshift/assisted-installer)
+will be enhanced accordingly to take care of this step, and to ensure it
+will be properly monitored during the installation process.
```
IMO it's not a nice-to-have feature but a critical step to make the installation workflow more robust and properly monitored. 
Copying 40Gb+ of data takes a good amount of time on a baremetal instance (up to 1hr as per the current testing), and several things can go wrong during the process, with the potential risk of leaving the data uncompleted or corrupted (as already experienced in the 4.19 DP release) - and thus the user should get properly informed that everything went fine (or if some problem happened, to avoid having later issues during the joining process, usually very painful to troubleshoot).
During the installation phase the Assisted Installer is the current eligible actor, on each node, responsible to carry on all the required task (such as writing the RHCOS image), and monitoring it's already supported via communication with the Assisted Service, which could be used by the user (via openshift-install wait-for command, or via the local Assisted UI) for that. And in this way we could also ensure that, prior of rebooting a node to let it join the control plane, all the required data (for making it working) will be effectively available.

andfasano commented:
/remove-lifecycle stale

rphillips commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,571 @@
+will be enhanced accordingly to take care of this step, and to ensure it
+will be properly monitored during the installation process.
+
+A new systemd service will be injected via MCO to run the registry container
```
An operator leveraging the static pod controller would be preferable here. This would be more in line with what the etcd and kube-apiserver operators do.

rphillips commented:
There are ramifications for having the registry on the master nodes. I see in previous comments this feature may be targeting bare metal, but if we want to target cloud machines, then we may have issues.

Network writes and disk reads can go against a cloud machines IOPS quotas, and could cause IO pauses issued from the provider's hypervisor. So the effect would show master nodes as down or not reachable. 

There is a case to move the registries to Worker nodes or required Infra nodes.

zaneb commented:
> There are ramifications for having the registry on the master nodes. I see in previous comments this feature may be targeting bare metal, but if we want to target cloud machines, then we may have issues.

@rphillips this is purely for on-prem. This feature requires setup at install time, and the agent installer does not support any cloud platforms.

andfasano committed:
update proposal with revised API, conditions model and implementation details

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+installation. The Appliance builder will be enhanced with a new 
+`build live-iso` command capable to generate an ISO artifact.
+Since the extended ISO will contain not only a specific set of OCP release images
+- but also additional images related to OLM operators (and other support images),
```
Nit - the '-' at the beginning of the line causes a bullet to be generated. Should change to a comma.

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+* The `status.releases.name` is the identifier of the release bundle
+* The `status.releases.image` is the OCP release image pullspec by digest for
+  the current release bundle
+* The `status.conditions` field will be used to keep track of the activity
```
Should this be` status.releases[].condition`?

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+      lastTransitionTime: "2024-11-01T07:00:00Z"
+  releases:
+    - name: ocp-release-bundle-4.18.0-x86_64
+      image: example.com/example/openshift-release-dev@sha256:aa8795f7932441b30bb8bcfbbf05912875383fad1f2b3be08a22ec148d6860ff
```
Should use real value of `api-int` here instead of `example.com` to make it clearer.

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+        message: "Release ocp-release-bundle-4.18.0-x86_64 is currently available"
+        lastTransitionTime: "2024-11-01T07:00:00Z"
+    - name: ocp-release-bundle-4.19.0-x86_64
+      image: example.com/example/openshift-release-dev@sha256:d98795f7932441b30bb8bcfbbf05912875383fad1f2b3be08a22ec148d68607f
```
Same comment about using `api-int`

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
````diff
@@ -0,0 +1,599 @@
+            lastTransitionTime: "2024-12-01T08:04:21Z"
+```
+
+* The `internalReleaseImage.status.releases` field will be used to track the status of the release
````
The format is actually `status.internalReleaseImage.XXX` not `internalReleaseImage.status.XXX`

andfasano replied:
ah good catch

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+
+## Upgrade / Downgrade Strategy
+
+N/A
```
Seems that there should be some description in this section since the doc covers upgrade/downgrades, or maybe just refer to the relevant section.

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+This ISO will be prepared and built using the [OpenShift Appliance Builder](https://github.com/openshift/appliance),
+a command line utility for building a disk image to orchestrate an OpenShift
+installation. The Appliance builder will be enhanced with a new 
+`build live-iso` command capable to generate an ISO artifact.
```
```suggestion
`build live-iso` command capable for generating an ISO artifact.
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+1. The user runs the `oc adm node-image create` command from within the
+   disconnected environment (with optionally some additional node
+   configuration).
+2. The user attaches the newly generate node ISO to the target machine(s),
```
```suggestion
2. The user attaches the newly generated node ISO to the target machine(s),
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+into the control planes nodes. In particular, it will be used for:
+
+* Triggering a new release image copy task - as a preliminary upgrade step
+* Delete from disk a release image version not anymore in use - as an optional
```
```suggestion
* Delete from disk a release image version no longer in use - as an optional
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+* `Removing`.  When true, it means that a release deletion is in progress on the current node, and not yet completed.
+* `Degraded`. When true, it means something has gone wrong in the current node.
+
+Additionaly, a new `StateProgress` enum value will be added to the MachineConfigNode conditions type:
```
```suggestion
Additionally, a new `StateProgress` enum value will be added to the MachineConfigNode conditions type:
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+
+##### Removal of a release from the InternalReleaseImage resource
+
+The user could remove a release payload not anymore in use by editing the
```
```suggestion
The user could remove a release payload no longer in use by editing the
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+required.
+Assisted Service will apply a pre-flight validation to ensure that the
+control plane nodes will meet the new storage requirement. Additionally, it
+should be properly documented in the pre-requistes section.
```
```suggestion
should be properly documented in the prerequistes section.
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+  limited set of OLM operators. Including all the OLM operators is not
+  feasible, and the current selected set may not be sufficient and/or
+  effectively required by all the users.
+* Due the size, some hardware vendors may have difficulties into loading
```
```suggestion
* Due to the size, some hardware vendors may have difficulties in loading
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+* Due the size, some hardware vendors may have difficulties into loading
+  such ISO, so a more extended testing may be required.
+* The lifecycle of some OLM operators is different from the OCP one,
+  thus assembling an extended ISO may be challenging due the different release
```
```suggestion
  thus assembling an extended ISO may be challenging due to the different release
```

bfournie commented on enhancements/agent-installer/simplified-registry-less-cluster-operations.md:
```diff
@@ -0,0 +1,599 @@
+  post-upgrade step
+* Opt-out from the feature (by deleting the resource)
+
+This is an example of how the CR will look like:
```
```suggestion
This is an example of what the CR will look like:
```

andfasano committed:
more grammar fixes

andfasano committed:
added missing OKE section
